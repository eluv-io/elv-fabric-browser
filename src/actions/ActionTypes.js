const ActionTypes = {
  accessGroups: {
    list: "LIST_ACCESS_GROUPS",
  },
  accounts: {
    get: {
      currentAccountAddress: "GET_CURRENT_ACCOUNT_ADDRESS"
    }
  },
  content: {
    libraries: {
      list: "LIST_LIBRARIES",
      get: "GET_LIBRARY",
      groups: "GET_LIBRARY_GROUPS",
      types: "GET_LIBRARY_TYPES"
    },
    objects: {
      list: "LIST_OBJECTS",
      get: "GET_OBJECTS",
      versions: "GET_OBJECT_VERSIONS"
    },
    types: {
      get: "GET_TYPE",
      list: "LIST_TYPES"
    }
  },
  contracts: {
    list: "LIST_CONTRACTS",
    compile: "COMPILE_CONTRACTS",
    deployed: {
      list: "LIST_DEPLOYED_CONTRACTS",
      balance: "GET_CONTRACT_BALANCE",
      events: "GET_CONTRACT_EVENTS",
      clearEvents: "CLEAR_CONTRACT_EVENTS",
      call: "CALL_CONTRACT_METHOD",
    },
    error: {
      compile: "ERROR_CONTRACT_COMPILATION"
    }
  },
  logs: {
    clear: "CLEAR_BLOCKCHAIN_LOGS",
    list: "LIST_BLOCKCHAIN_LOGS"
  },
  notifications: {
    error: "SET_ERROR_MESSAGE",
    notification: "SET_NOTIFICATION_MESSAGE",
    clear: "CLEAR_NOTIFICATIONS"
  },
  requests: {
    status: {
      submitted: "REQUEST_SUBMITTED",
      completed: "REQUEST_COMPLETED",
      error: "REQUEST_ERROR"
    }
  },
  routes: {
    path: "ROUTING_FRAME_PATH",
    synchronize: "START_ROUTE_SYNCHRONIZATION"
  }
};

export default ActionTypes;
