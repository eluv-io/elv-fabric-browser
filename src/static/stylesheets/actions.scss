@import "variables/mixins";
@import "variables/colors";

.actions-container {
  display: flex;

  :first-child {
    &.action,
    &.action-loading {
      margin-left: 0;
    }
  }

  :last-child {
    &.action,
    &.action-loading {
      margin-right: 0;
    }
  }

  &.loading {
    justify-content: center;
    padding: 17px;
  }

  .action,
  .action-loading {
    @include no-select;
    align-items: center;
    box-sizing: border-box;
    display: flex;
    justify-content: center;
    margin: 10px 10px 10px 0;
    min-width: 100px;
  }

  .action {
    background: $color-mediumblue;
    border: 0;
    color: $color-text-white;
    cursor: pointer;
    font-size: 14px;
    height: 2.25em;
    justify-content: center;
    min-width: 100px;
    padding: 7px 15px;

    &:hover {
      background: $color-lightblue;
    }

    &.secondary {
      background: $color-action-secondary;
      color: $color-text-dark;

      &:hover {
        background: $color-lightgray;
      }
    }

    &.tertiary {
      background: $color-bg-white;
      border: 1px solid $color-action-secondary;
      color: $color-text-dark;

      &:hover {
        border: 1px solid $color-darkgray;
      }
    }

    &.action-compact {
      font-size: 12px;
      height: auto;
      margin: 0;
      min-width: 75px;
      padding: 5px;
    }

    &.action-wide {
      min-width: 120px;
    }
  }

  .action-loading {
    color: $color-midgray;
    margin: 0;
  }
}
