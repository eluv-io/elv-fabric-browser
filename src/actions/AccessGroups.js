import ActionTypes from "./ActionTypes";
import Fabric from "../clients/Fabric";
import {SetNotificationMessage} from "./Notifications";
import { FormatAddress } from "../utils/Helpers";
import {WithCancel} from "../utils/Cancelable";

export const ListAccessGroups = ({params}) => {
  return async (dispatch) => {
    const {accessGroups, count} = await WithCancel(
      params.cancelable,
      async () => await Fabric.AccessGroups({params})
    );

    dispatch({
      type: ActionTypes.accessGroups.list,
      accessGroups,
      count
    });
  };
};

export const GetAccessGroup = ({contractAddress}) => {
  return async (dispatch) => {
    const accessGroup = await Fabric.GetAccessGroup({contractAddress});

    dispatch({
      type: ActionTypes.accessGroups.get,
      contractAddress,
      accessGroup
    });
  };
};

export const SaveAccessGroup = ({address}) => {
  return async (dispatch) => {
    if(address) {
      dispatch(SetNotificationMessage({
        message: "Access group successfully updated",
        redirect: true
      }));
    } else {
      // New access group - deploy contract
      address = await Fabric.CreateAccessGroup();

      dispatch(SetNotificationMessage({
        message: "Access group successfully created",
        redirect: true
      }));
    }

    return FormatAddress(address);
  };
};

export const ListAccessGroupMembers = ({contractAddress, params}) => {
  return async (dispatch) => {
    const {members, count} = await WithCancel(
      params.cancelable,
      async () => await Fabric.AccessGroupMembers({contractAddress, params})
    );

    dispatch({
      type: ActionTypes.accessGroups.members.list,
      contractAddress,
      members,
      count
    });
  };
};

// Determine diff between new and old members and add/remove access as necessary
const UpdateMembers = async ({contractAddress, members, originalMembers}) => {
  const newMembers = Object.values(members).filter(member => !member.manager).map(member => member.address);
  const oldMembers = Object.values(originalMembers).filter(member => !member.manager).map(member => member.address);

  const newManagers = Object.values(members).filter(member => member.manager).map(member => member.address);
  const oldManagers = Object.values(originalMembers).filter(member => member.manager).map(member => member.address);

  const membersToAdd = newMembers.filter(x => !oldMembers.includes(x));
  for(const memberAddress of membersToAdd) {
    await Fabric.AddAccessGroupMember({contractAddress, memberAddress});
  }

  const membersToRemove = oldMembers.filter(x => !newMembers.includes(x));
  for(const memberAddress of membersToRemove) {
    await Fabric.RemoveAccessGroupMember({contractAddress, memberAddress});
  }

  const managersToAdd = newManagers.filter(x => !oldManagers.includes(x));
  for(const memberAddress of managersToAdd) {
    await Fabric.AddAccessGroupManager({contractAddress, memberAddress});
  }

  const managersToRemove = oldManagers.filter(x => !newManagers.includes(x));
  for(const memberAddress of managersToRemove) {
    await Fabric.RemoveAccessGroupManager({contractAddress, memberAddress});
  }
};

export const UpdateAccessGroupMembers = ({address, members, originalMembers}) => {
  return async (dispatch) => {
    const accessGroup = await Fabric.GetAccessGroup({contractAddress: address});

    if(!accessGroup) { throw Error("Access group not found"); }

    await UpdateMembers({contractAddress: address, members, originalMembers});


    dispatch(SetNotificationMessage({
      message: "Access group members successfully updated",
      redirect: true
    }));
  };
};

export const RemoveAccessGroup = ({address}) => {
  return async (dispatch) => {
    await Fabric.DeleteAccessGroup({address});

    dispatch(SetNotificationMessage({
      message: "Access group successfully deleted",
      redirect: true
    }));
  };
};
