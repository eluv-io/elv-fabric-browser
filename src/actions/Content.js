import ActionTypes from "./ActionTypes";
import Fabric from "../clients/Fabric";
import { SetNotificationMessage } from "./Notifications";
import { ParseInputJson } from "../utils/Input";
import {FormatAddress} from "../utils/Helpers";
import {ToList} from "../utils/TypeSchema";
import {DownloadFromUrl, FileInfo} from "../utils/Files";
import Path from "path";

export const ListContentLibraries = () => {
  return async (dispatch) => {
    let libraries = await Fabric.ListContentLibraries();

    // Exclude special content space library
    delete libraries[Fabric.contentSpaceLibraryId];

    dispatch({
      type: ActionTypes.content.libraries.list,
      libraries
    });
  };
};

export const GetContentLibrary = ({libraryId}) => {
  return async (dispatch) => {
    const library = await Fabric.GetContentLibrary({libraryId});

    dispatch({
      type: ActionTypes.content.libraries.get,
      libraryId: libraryId,
      library
    });
  };
};

export const ListLibraryContentTypes = ({libraryId}) => {
  return async (dispatch) => {
    const types = await Fabric.ListLibraryContentTypes({libraryId});

    dispatch({
      type: ActionTypes.content.libraries.types,
      libraryId: libraryId,
      types
    });
  };
};

export const ListContentLibraryGroups = ({libraryId}) => {
  return async (dispatch) => {
    // Content space does not have groups
    if(libraryId === Fabric.contentSpaceLibraryId) { return; }

    const groups = await Fabric.GetContentLibraryGroups({libraryId});

    dispatch({
      type: ActionTypes.content.libraries.groups,
      libraryId,
      groups
    });
  };
};

export const CreateContentLibrary = ({name, description, publicMetadata, privateMetadata, image}) => {
  return async (dispatch) => {
    const libraryId = await Fabric.CreateContentLibrary({
      name,
      description,
      publicMetadata: ParseInputJson(publicMetadata),
      privateMetadata: ParseInputJson(privateMetadata)
    });

    if(image) {
      await Fabric.SetContentLibraryImage({
        libraryId,
        image: await new Response(image).blob()
      });
    }

    dispatch(SetNotificationMessage({
      message: "Successfully created content library '" + name + "'",
      redirect: true
    }));

    return libraryId;
  };
};

export const UpdateContentLibrary = ({
  libraryId,
  name,
  description,
  publicMetadata,
  privateMetadata,
  image
}) => {
  return async (dispatch) => {
    publicMetadata = ParseInputJson(publicMetadata);
    publicMetadata["eluv.name"] = name;
    publicMetadata["eluv.description"] = description;

    await Fabric.ReplacePublicLibraryMetadata({
      libraryId,
      metadata: publicMetadata
    });

    const libraryObjectId = libraryId.replace("ilib", "iq__");
    await Fabric.EditAndFinalizeContentObject({
      libraryId,
      objectId: libraryObjectId,
      todo: async (writeToken) => {
        await Fabric.ReplaceMetadata({
          libraryId,
          objectId: libraryObjectId,
          writeToken,
          metadata: ParseInputJson(privateMetadata)
        });
      }
    });

    if(image) {
      await Fabric.SetContentLibraryImage({
        libraryId,
        image: await new Response(image).blob()
      });
    }

    dispatch(SetNotificationMessage({
      message: "Successfully updated content library '" + name + "'",
      redirect: true
    }));
  };
};

export const DeleteContentLibrary = ({ libraryId }) => {
  return async (dispatch) => {
    await Fabric.DeleteContentLibrary({libraryId});

    dispatch(SetNotificationMessage({
      message: "Successfully deleted content library",
      redirect: true
    }));
  };
};

export const SetLibraryContentTypes = ({libraryId, typeIds=[]}) => {
  return async (dispatch) => {
    const currentTypeIds = Object.values(await Fabric.ListLibraryContentTypes({libraryId}))
      .map(type => type.id);

    const idsToAdd = typeIds.filter(id => !currentTypeIds.includes(id));

    if(idsToAdd.length > 0) {
      const contentTypes = Object.values(await Fabric.ListContentTypes({}));
      for (const typeId of idsToAdd) {
        // When adding a content type, check for custom contract
        const type = contentTypes.find(type => type.id === typeId);
        const customContractAddress = type.meta.custom_contract && type.meta.custom_contract.address;
        await Fabric.AddLibraryContentType({libraryId, typeId, customContractAddress});
      }
    }

    const idsToRemove = currentTypeIds.filter(id => !typeIds.includes(id));

    for(const typeId of idsToRemove) {
      await Fabric.RemoveLibraryContentType({libraryId, typeId});
    }

    dispatch(SetNotificationMessage({
      message: "Successfully updated content library types",
      redirect: true
    }));
  };
};

export const UpdateContentLibraryGroups = ({libraryId, groups, originalGroups}) => {
  return async (dispatch) => {
    for(const groupType of Object.keys(groups)) {
      const oldGroupAddresses = originalGroups[groupType].map(group => FormatAddress(group.address));
      const newGroupAddresses = groups[groupType].map(group => FormatAddress(group.address));

      // Remove groups in original groups but not in new groups
      const toRemove = oldGroupAddresses.filter(address => !newGroupAddresses.includes(address));
      for(const address of toRemove) {
        await Fabric.RemoveContentLibraryGroup({libraryId, address, groupType});
      }

      // Add groups in new groups but not in original groups
      const toAdd = newGroupAddresses.filter(address => !oldGroupAddresses.includes(address));
      for(const address of toAdd) {
        await Fabric.AddContentLibraryGroup({libraryId, address, groupType});
      }
    }

    dispatch(SetNotificationMessage({
      message: "Successfully updated library groups",
      redirect: true
    }));
  };
};

export const ListContentObjects = ({libraryId}) => {
  return async (dispatch) => {
    const objects = await Fabric.ListContentObjects({libraryId});

    dispatch({
      type: ActionTypes.content.objects.list,
      objects
    });
  };
};

export const GetContentObject = ({libraryId, objectId}) => {
  return async (dispatch) => {
    const object = await Fabric.GetContentObject({libraryId, objectId});

    dispatch({
      type: ActionTypes.content.objects.get,
      objectId: object.id,
      object
    });
  };
};

export const GetContentObjectVersions = ({libraryId, objectId}) => {
  return async (dispatch) => {
    const versions = await Fabric.GetContentObjectVersions({libraryId, objectId});

    dispatch({
      type: ActionTypes.content.objects.versions,
      objectId,
      versions
    });
  };
};

export const GetContentObjectPermissions = async ({libraryId, objectId}) => {
  return await Fabric.GetContentObjectPermissions({libraryId, objectId});
};

export const PublishContentObject = async ({objectId}) => {
  return await Fabric.PublishContentObject({objectId});
};

export const ReviewContentObject = async ({libraryId, objectId, approve, note}) => {
  await Fabric.ReviewContentObject({libraryId, objectId, approve, note});

  const currentAccountAddress = await Fabric.CurrentAccountAddress();

  await Fabric.EditAndFinalizeContentObject({
    libraryId,
    objectId,
    todo: async (writeToken) => {
      await Fabric.MergeMetadata({
        libraryId,
        writeToken,
        metadata: {
          "eluv.reviewer": currentAccountAddress,
          "eluv.reviewNote": note
        }
      });
    }
  });
};

export const CreateContentObject = ({libraryId, name, description, type, metadata}) => {
  return async (dispatch) => {
    metadata = ParseInputJson(metadata);
    metadata["eluv.name"] = name;
    metadata["eluv.description"] = description;

    const objectInfo = await Fabric.CreateAndFinalizeContentObject({
      libraryId,
      type,
      metadata,
    });

    dispatch(SetNotificationMessage({
      message: "Successfully created content object",
      redirect: true
    }));

    return objectInfo.id;
  };
};

export const DeleteContentObject = ({ libraryId, objectId }) => {
  return async (dispatch) => {
    await Fabric.DeleteContentObject({libraryId, objectId});

    dispatch(SetNotificationMessage({
      message: "Successfully deleted content object",
      redirect: true
    }));
  };
};

export const DeleteContentVersion = ({ libraryId, objectId, versionHash }) => {
  return async (dispatch) => {
    await Fabric.DeleteContentVersion({libraryId, objectId, versionHash});

    dispatch(SetNotificationMessage({
      message: "Successfully deleted content version",
      redirect: true
    }));
  };
};

export const UpdateContentObject = ({libraryId, objectId, name, description, type, metadata}) => {
  return async (dispatch) => {
    let contentDraft = await Fabric.EditContentObject({
      libraryId,
      objectId,
      options: {
        type
      }
    });

    metadata = ParseInputJson(metadata);
    metadata["eluv.name"] = name;
    metadata["eluv.description"] = description;

    await Fabric.ReplaceMetadata({
      libraryId,
      objectId,
      writeToken: contentDraft.write_token,
      metadata
    });

    await Fabric.FinalizeContentObject({
      libraryId,
      objectId,
      writeToken: contentDraft.write_token
    });

    dispatch(SetNotificationMessage({
      message: "Successfully updated content object",
      redirect: true
    }));
  };
};

export const CreateContentType = ({name, description, metadata, bitcode}) => {
  return async (dispatch) => {
    bitcode = await new Response(bitcode).blob();

    const objectId = await Fabric.CreateContentType({
      name,
      description,
      metadata: ParseInputJson(metadata),
      bitcode
    });

    dispatch(SetNotificationMessage({
      message: "Successfully created content type",
      redirect: true
    }));

    return objectId;
  };
};

export const ListContentTypes = ({latestOnly=true}) => {
  return async (dispatch) => {
    dispatch({
      type: ActionTypes.content.types.list,
      types: await Fabric.ListContentTypes({latestOnly})
    });
  };
};

export const GetContentType = ({versionHash}) => {
  return async (dispatch) => {
    dispatch({
      type: ActionTypes.content.types.get,
      contentType: await Fabric.GetContentType({versionHash})
    });
  };
};

export const UploadParts = ({libraryId, objectId, files, callback, encrypt}) => {
  return async (dispatch) => {
    let contentDraft = await Fabric.EditContentObject({libraryId, objectId});

    await Promise.all(Array.from(files).map(async file => {
      const data = await new Response(file).blob();

      let partCallback;
      if(callback) {
        partCallback = ({uploaded, total}) => callback({uploaded, total, filename: file.name});
      }

      await Fabric.UploadPart({
        libraryId,
        objectId,
        writeToken: contentDraft.write_token,
        data,
        chunkSize: 10000000,
        callback: partCallback,
        encrypted: encrypt
      });
    }));

    await Fabric.FinalizeContentObject({
      libraryId,
      objectId,
      writeToken: contentDraft.write_token
    });

    const partsText = files.length > 1 ? "parts" : "part";

    dispatch(SetNotificationMessage({
      message: "Successfully uploaded " + partsText,
      redirect: true
    }));
  };
};

export const DownloadPart = ({libraryId, objectId, versionHash, partHash, callback}) => {
  return async (dispatch) => {
    let blob = await Fabric.DownloadPart({libraryId, objectId, versionHash, partHash, encrypted: false});
    let url = window.URL.createObjectURL(blob);

    await callback(url);
  };
};

export const UploadFiles = ({libraryId, objectId, path, fileList}) => {
  return async (dispatch) => {
    await Fabric.EditAndFinalizeContentObject({
      libraryId,
      objectId,
      todo: async (writeToken) => {
        const fileInfo = await FileInfo(path, fileList);

        await Fabric.UploadFiles({libraryId, objectId, writeToken, fileInfo});
      }
    });

    dispatch(SetNotificationMessage({
      message: "Successfully uploaded files"
    }));
  };
};

export const DownloadFile = ({libraryId, objectId, versionHash, filePath}) => {
  return async (dispatch) => {
    let blob = await Fabric.DownloadFile({libraryId, objectId, versionHash, filePath});
    let url = window.URL.createObjectURL(blob);

    await DownloadFromUrl(url, Path.basename(filePath));
  };
};

export const FileUrl = ({libraryId, objectId, versionHash, filePath}) => {
  return async (dispatch) => {
    return await Fabric.FileUrl({libraryId, objectId, versionHash, filePath});
  };
};

const CollectMetadata = async ({libraryId, writeToken, schema, fields}) => {
  let metadata = {};

  for(const entry of schema) {
    switch(entry.type) {
      case "label":
      case "attachedFile":
        break;
      case "file":
        // Freshly uploaded files will be a FileList
        // previously uploaded files will either be a string or a list of strings
        if(Array.isArray(fields[entry.key]) || typeof fields[entry.key] === "string") {
          metadata[entry.key] = fields[entry.key];
          break;
        }

        const files = Array.from(fields[entry.key]);
        let partResponses = [];

        for (const file of files) {
          const data = await new Response(file).blob();

          partResponses.push(
            await Fabric.UploadPart({
              libraryId,
              writeToken,
              data,
              encrypted: !!(entry.encrypted)
            })
          );
        }

        if (entry.multiple) {
          metadata[entry.key] = partResponses.map(partResponse => partResponse.part.hash);
        } else {
          metadata[entry.key] = partResponses.length > 0 ? partResponses[0].part.hash : "";
        }

        break;

      case "json":
        metadata[entry.key] = ParseInputJson(fields[entry.key]);
        break;

      case "list":
        metadata[entry.key] = ToList(fields[entry.key]);
        break;
      case "object":
        metadata[entry.key] = await CollectMetadata({
          libraryId,
          writeToken,
          schema: entry.fields,
          fields: fields[entry.key]
        });
        break;

      default:
        metadata[entry.key] = fields[entry.key];
    }
  }

  return metadata;
};

export const CreateFromContentTypeSchema = ({libraryId, type, metadata, accessCharge, schema, fields}) => {
  return async (dispatch) => {
    const createResponse = await Fabric.CreateAndFinalizeContentObject({
      libraryId,
      type,
      todo: async (writeToken) => {
        await Fabric.ReplaceMetadata({
          libraryId,
          writeToken,
          metadata: {
            ...ParseInputJson(metadata),
            ...(await CollectMetadata({libraryId, writeToken, schema, fields})),
          }
        });
      }
    });

    if(accessCharge > 0) {
      await Fabric.SetAccessCharge({objectId: createResponse.id, accessCharge});
    }

    dispatch(SetNotificationMessage({
      message: "Successfully created content",
      redirect: true
    }));
  };
};

export const UpdateFromContentTypeSchema = ({libraryId, objectId, metadata, accessCharge, schema, fields}) => {
  return async (dispatch) => {
    await Fabric.EditAndFinalizeContentObject({
      libraryId,
      objectId,
      todo: async (writeToken) => {
        await Fabric.ReplaceMetadata({
          libraryId,
          writeToken,
          metadata: {
            ...ParseInputJson(metadata),
            ...(await CollectMetadata({libraryId, writeToken, schema, fields}))
          }
        });
      }
    });

    await Fabric.SetAccessCharge({objectId: objectId, accessCharge});

    dispatch(SetNotificationMessage({
      message: "Successfully updated content",
      redirect: true
    }));
  };
};

export const AddApp = ({libraryId, objectId, role, file}) => {
  return async (dispatch) => {
    await Fabric.EditAndFinalizeContentObject({
      libraryId,
      objectId,
      todo: async (writeToken) => {
        // Override given filename
        const fileName = `${role}App.html`;
        file[0].overrideName = fileName;

        const fileInfo = await FileInfo("/", file);

        await Fabric.UploadFiles({
          libraryId,
          objectId,
          writeToken,
          fileInfo
        });

        await Fabric.ReplaceMetadata({
          libraryId,
          objectId,
          writeToken,
          metadataSubtree: `eluv.${role}App`,
          metadata: fileName
        });
      }
    });

    dispatch(SetNotificationMessage({
      message: "Successfully added " + role + " app",
      redirect: true
    }));
  };
};

export const RemoveApp = ({libraryId, objectId, role}) => {
  return async (dispatch) => {
    await Fabric.EditAndFinalizeContentObject({
      libraryId,
      objectId,
      todo: async (writeToken) => {
        await Fabric.DeleteMetadata({
          libraryId,
          objectId,
          writeToken,
          metadataSubtree: `eluv.${role}App`
        });
      }
    });

    dispatch(SetNotificationMessage({
      message: "Successfully removed " + role + " app",
      redirect: true
    }));
  };
};
