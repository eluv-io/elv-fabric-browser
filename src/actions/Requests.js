import ActionTypes from "./ActionTypes";
import { SetErrorMessage } from "./Notifications";
import Id from "../utils/Id";
import { Wait } from "../utils/Helpers";

// Wrap actions with requests to keep request state updated automatically
// Returns a unique ID corresponding to the request state
export const WrapRequest = (
  {
    dispatch,
    domain,
    action,
    modal=false,
    todo
  }
) => {
  let requestId = Id.next();

  Request({
    requestId,
    dispatch,
    domain,
    action,
    modal,
    todo
  });

  return requestId;
};

const Request = async (
  {
    requestId,
    dispatch,
    domain,
    action,
    modal,
    todo
  }
) => {
  dispatch(RequestSubmitted(requestId, domain, action));

  try {
    // Simulate loading time
    //await Wait(500);

    await todo();

    dispatch(RequestCompleted(requestId, domain, action));
  } catch(error) {
    let errorMessage = error;
    if(typeof error !== "string") {
      // Error object -- actual message may be present in one of several different keys
      errorMessage = error.message ||
        error.statusText ||
        (error.responseText && JSON.parse(error.responseText).error.message) ||
        error;
    }

    dispatch(RequestError(requestId, domain, action, errorMessage));

    // Modals will handle their own errors
    if(!modal) {
      dispatch(SetErrorMessage({message: errorMessage}));
    }
  }
};

const RequestSubmitted = (requestId, domain, action) => {
  return {
    type: ActionTypes.request.request.status.submitted,
    requestId: requestId,
    domain: domain,
    action: action
  };
};

const RequestCompleted = (requestId, domain, action) => {
  return {
    type: ActionTypes.request.request.status.completed,
    requestId: requestId,
    domain: domain,
    action: action
  };
};

const RequestError = (requestId, domain, action, error_message) => {
  return {
    type: ActionTypes.request.request.status.error,
    requestId: requestId,
    domain: domain,
    action: action,
    error_message: error_message
  };
};
