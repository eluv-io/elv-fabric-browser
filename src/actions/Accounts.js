import ActionTypes from "./ActionTypes";
import Fabric from "../clients/Fabric";
import { SetNotificationMessage } from "./Notifications";
import { WrapRequest } from "./Requests";

export const ListAccounts = () => {
  return (dispatch) => {
    return WrapRequest({
      dispatch: dispatch,
      domain: "accounts",
      action: "listAccounts",
      todo: async () => {
        /*
        let accountAddresses = await Ethereum.ListAccounts();

        let accounts = {};
        accountAddresses.forEach(
          (address) => { accounts[address] = {}; }
        );

        dispatch({
          type: ActionTypes.request.accounts.completed.list.accounts,
          accounts: accounts
        });
        */
      }
    });
  };
};

export const CreateAccount = ({password}) => {
  return (dispatch) => {
    return WrapRequest({
      dispatch: dispatch,
      domain: "accounts",
      action: "createAccount",
      todo: async () => {
        /*
        let accountAddress = await Ethereum.CreateAccount({ password: password });

        dispatch(SetNotificationMessage({
          message: "Successfully created account '" + accountAddress + "'",
          redirect: true
        }));
        */
      }
    });
  };
};

export const GetAccountInfo = ({accountAddress}) => {
  return (dispatch) => {
    return WrapRequest({
      dispatch: dispatch,
      domain: "accounts",
      action: "getAccountInfo",
      todo: async () => {
        /*
        let libraryResponse = await Fabric.GetByName({name: "ElvAccountInfo" });
        let accountObject = await Fabric.GetObjectByName({name: accountAddress});
        accountObject.libraryId = libraryResponse.target;

        accountObject.accountBalance = await Ethereum.Balance(accountAddress);
        accountObject.accountTransactionCount = await Ethereum.TransactionCount(accountAddress);

        dispatch({
          type: ActionTypes.request.accounts.completed.list.account,
          accountAddress,
          accountInfo: accountObject
        });
        */
      }
    });
  };
};

export const UpdateAccountInfo = ({
  accountAddress,
  name,
  profileImageFile,
  bio
}) => {
  return (dispatch) => {
    return WrapRequest({
      dispatch: dispatch,
      domain: "accounts",
      action: "updateAccountInfo",
      todo: async () => {
        // TODO: This should be replaced by some sort of first class fabric API for storing account info
        // For now, this library needs to be created and named manually
        let accountLibraryInfo = await Fabric.GetByName({name: "ElvAccountInfo"});
        if(!accountLibraryInfo) {
          console.log("Create the account info library and use naming service to name it ElvAccountInfo");
          throw Error();
        }
        let libraryId = accountLibraryInfo.target;

        let accountInfoObject = await Fabric.GetObjectByName({name: accountAddress});

        let writeToken;

        if(accountInfoObject) {
          // Edit existing info object
          let editResponse = await Fabric.EditContentObject({
            libraryId,
            objectId: accountInfoObject.objectId
          });

          writeToken = editResponse.write_token;

          await Fabric.MergeMetadata({
            libraryId,
            objectId: accountInfoObject.objectId,
            writeToken,
            metadata: {
              name,
              bio
            }
          });
        } else {
          // Create new info object
          let createResponse = await Fabric.CreateContentObject({
            libraryId,
            name: "Account Info: " + accountAddress,
            metadata: {
              name,
              bio
            }
          });
          writeToken = createResponse.write_token;
        }

        if(profileImageFile) {
          const imageData = await new Response(profileImageFile).blob();
          const uploadResponse = await Fabric.UploadPart({
            libraryId,
            objectId: accountInfoObject.objectId,
            writeToken,
            data: imageData
          });

          await Fabric.MergeMetadata({
            libraryId,
            objectId: accountInfoObject.objectId,
            writeToken,
            metadata: {
              profileImageHash: uploadResponse.part.hash
            }
          });
        }

        // TODO: IMAGE

        let finalizeResponse = await Fabric.FinalizeContentObject({
          libraryId,
          objectId: accountInfoObject.objectId,
          writeToken
        });

        await Fabric.SetObjectByName({
          name: accountAddress,
          libraryId,
          objectId: finalizeResponse.id
        });

        dispatch(SetNotificationMessage({
          message: "Successfully updated account info",
          redirect: true
        }));
      }
    });
  };
};









