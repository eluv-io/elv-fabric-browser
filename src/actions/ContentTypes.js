import ActionTypes from "./ActionTypes";
import Fabric from "../clients/Fabric";
import { SetNotificationMessage } from "./Notifications";
import { WrapRequest } from "./Requests";

export const ListContentTypes = () => {
  return (dispatch) => {
    return WrapRequest({
      dispatch: dispatch,
      domain: "content",
      action: "listContentTypes",
      todo: (async () => {
        /*
          let libraryIds = await Fabric.ListContentLibraries();

          let result = await ParseContentLibraries({libraryIds: libraryIds});

          dispatch({
            type: ActionTypes.request.content.completed.list.all,
            contentLibraries: result.contentLibraries
          });
        */
      })
    });
  };
};
