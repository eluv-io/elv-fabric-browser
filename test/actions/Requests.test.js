import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { WrapRequest } from "../../src/actions/Requests";
import {AwaitRequest} from "../utils";

const mockStore = configureStore([ thunk ]);
const store = mockStore();

describe("Request wrapper", () => {
  beforeEach(() => {
    store.clearActions();
  });

  test("Successful request", async () => {
    const requestId = WrapRequest({
      dispatch: store.dispatch,
      domain: "test-domain",
      action: "test-action",
      todo: () => {
        return "Successful action";
      }
    });

    expect(typeof requestId).toBe("number");

    expect(await AwaitRequest({ store, requestId})).toBe(true);

    expect(store.getActions()).toMatchSnapshot();
  });

  test("Failed request", async () => {
    const todo = () => { throw Error("Invalid request"); };
    const requestId = WrapRequest({
      dispatch: store.dispatch,
      domain: "test-domain",
      action: "test-action",
      todo
    });

    expect(typeof requestId).toBe("number");

    expect(await AwaitRequest({ store, requestId})).toBe(false);

    expect(store.getActions()).toMatchSnapshot();
  });
});
