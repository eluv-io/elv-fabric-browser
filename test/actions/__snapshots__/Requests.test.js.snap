// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request wrapper Failed request 1`] = `
Array [
  Object {
    "action": "test-action",
    "domain": "test-domain",
    "requestId": 2,
    "type": "REQUEST_SUBMITTED",
  },
  Object {
    "action": "test-action",
    "domain": "test-domain",
    "error_message": "Invalid request",
    "requestId": 2,
    "type": "REQUEST_ERROR",
  },
  Object {
    "message": "Invalid request",
    "redirect": false,
    "type": "SET_ERROR_MESSAGE",
  },
]
`;

exports[`Request wrapper Successful request 1`] = `
Array [
  Object {
    "action": "test-action",
    "domain": "test-domain",
    "requestId": 1,
    "type": "REQUEST_SUBMITTED",
  },
  Object {
    "action": "test-action",
    "domain": "test-domain",
    "requestId": 1,
    "type": "REQUEST_COMPLETED",
  },
]
`;
